pkg/querier/queryrange/results_cache.go

```diff
@@ -161,6 +166,10 @@ func (s resultsCache) Do(ctx context.Context, r Request) (Response, error) {
 		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
 
+	if s.shouldCache != nil && !s.shouldCache(r) {
```

Comment:
I would maybe create a separate quick middleware (:

You want to implement 

```go
// Middleware is a higher order Handler.
type Middleware interface {
	Wrap(Handler) Handler
}

```

Reply:
I checked the code and I found that this way makes the implementation much more complicated. Is this change worthwhile? And I think using the function in this pr is already very clear, is it really necessary to always follow the same pattern?

Reply:
``` go
type cacheChecker struct {
	resultsCache queryrange.Handler

	// cacheNextHandler is the next handler of results cache
	cacheNextHandler queryrange.Handler
}

func newCacheChecker(cacheNextHandler queryrange.Handler) queryrange.Middleware {
	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
		return &cacheChecker{
			resultsCache:     next,
			cacheNextHandler: cacheNextHandler,
		}
	})
}

func (c *cacheChecker) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
	if r.GetNoCache() {
		return c.cacheNextHandler.Do(ctx, r)
	}

	return c.resultsCache.Do(ctx, r)
}
```

Above is the example implementation for this middleware. But when initializing this middleware we need to explicitly pass the next handler of results cache. However, this is not doable I think because we can only get `Middleware` type instead of `Handler`. https://github.com/thanos-io/thanos/blob/master/pkg/queryfrontend/roundtrip.go#L93.

Sorry for the long comment. If I missed anything, please let me know!

Reply:
There is some boilerplate to it, yes so for simple if, let's try your solution then, thanks for trying!

Reply:
I would have suggested the middleware too, but I agree to keep it simple if the middleware is complicated, so 👍